Newsletters: add 104 (2020-07-01) (#427)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-07-01-newsletter.md

    resolve their preimages well before their timelocks expire.  In the
    long term, this is an incentive compatibility concern that
    protocol developers may want to keep in mind.

- **Proposed service for storing, relaying, and broadcasting presigned transactions:**
  Jacob Swambo sent a [request for comments][swambo rfc] to the Bitcoin-Dev
  mailing list about creating software and a protocol for allowing
  applications to store presigned transactions with third-party services
  for later broadcast.  The software could also perhaps determine when to
  broadcast the transactions based on certain conditions being met.  This
  could be useful for software such as [vaults][topic vaults] and
  [watchtowers][topic watchtowers].  Swambo asks anyone interested in
  the idea, especially those interested in using it with their protocol, to
  contact him.

## Recently transcribed talks and conversations

*[Bitcoin Transcripts][] is the home for transcripts of technical
Bitcoin presentations and discussions. In this monthly feature, we
highlight a selection of the transcripts from the previous month.*

- **Magical Bitcoin:** Alekos Filini presented at LA BitDevs on [Magical

>   Bitcoin][], an in-development set of tools and libraries for onchain

Reviewer: perhaps s/an in-development set of tools and libraries/a set of tools and libraries under development/

  visualizations. It is written in Rust and leverages the
  rust-miniscript library and the open source Esplora block explorer.
  ([transcript][mb ts], [video][mb vid])

- **Watchtowers and BOLT13:** Sergi Delgado appeared on [Potzblitz][] to
  discuss the latest state of [watchtower][topic watchtowers]
  development and a proposed watchtower protocol specification. He
  explored the various tradeoffs when designing a watchtower and the
  interplay between privacy requirements, accessibility, storage, and
  fees charged.  Delgado is working on the watchtower implementation
  [The Eye of Satoshi][teos] at Talaia Labs, which is aiming to be
  compliant with the proposed BOLT13 (see [Newsletter #75][news75
  watchtower bolt]).  Delgado also highlighted how watchtowers are
  becoming increasingly critical in multiple settings such as Bitcoin
  [vault][topic vaults] designs, statechains, and atomic swaps in addition to LN.
  ([transcript][watchtowers ts], [video][watchtowers vid],
  [slides][watchtowers slides])

- **CoinSwap:** Aviv Milner presented on CoinSwap at the Wasabi Research
  Club. He explained the property of covertness and how Chris Belcher’s

>   CoinSwap proposal (see [Newsletter #100][news100 coinswap]) provides

Reviewer: coinswap? here and below, per style guide (unless this is a specific variant = proper noun)

  [coinjoin][topic coinjoin] fail to do. Milner also went through the
  motivation for routing CoinSwaps, namely to address the concern of
  unwittingly entering into a CoinSwap with an adversary such a chain
  surveillance company.  ([transcript][coinswap ts], [video][coinswap
  vid])

- **Schnorr signatures and multisignatures:** [BIP340][] co-authors Tim
  Ruffing, Pieter Wuille, and Jonas Nick held a discussion at London
  Bitcoin Devs on the specification of [schnorr signatures][topic
  schnorr signatures].  This covered earlier ideas for implementing
  schnorr signatures in Bitcoin and what the co-authors thought the
  community should be concerned about when considering a possible future
  soft fork deployment. Wuille noted that he is most concerned about how
  schnorr is adopted and what is built using it.  The following day, Tim
  Ruffing presented on the challenges of multisignature and threshold
  signature schemes using schnorr signatures.  He's been working with
  collaborators on a speculative [MuSig][topic musig] signature scheme
  that only requires two rounds of interaction rather than three,
  without relying on zero knowledge proofs, which would greatly simplify

>   some threshold signing workflows.  ([Meeting transcript][socratic ts],

Reviewer 2: I would prefer to revert the change made to this Schnorr section as I think the original was stronger. First the speculative MuSig scheme doesn't just impact threshold signing workflows it also impact multisignature workflows. I also thought the original discussion of the added complexities on threshold over multisig was interesting.

Author: It's your call.  Would you mind if I give it another try to keep it short but add a link to https://bitcoinops.org//en/newsletters/2019/10/16/#the-quest-for-practical-threshold-schnorr-signatures where we previously described his concerns?

Reviewer 2: Sure that link is a good one. In that case "simplify some **multisignature and** threshold signing workflows"?

  [meeting video][socratic vid], [presentation transcript][ruffing ts],
  [presentation video][ruffing vid], [presentation slides][ruffing
  slides])

- **Sydney meetup discussion:** A number of Bitcoin and LN developers
  joined this Sydney meetup to discuss various topics. Ruben Somsen gave
  a short presentation on Succinct Atomic Swaps (SAS) (see [Newsletter
  #98][news98 sas]) and how it compares to Chris Belcher’s CoinSwap
  proposal.  Another topic was Bitcoin Core's reliance on DNS seeds to

>   find initial peers, whether that's an attack vector against newly

Reviewer: perhaps s/that's an/it represents a potential/

  started nodes, and how work by Matt Corallo and Antoine Riard on
  allowing [alternative transports][Bitcoin Core #18988] could help

>   mitigate any risk.  Finally, Lloyd Fournier discussed how

Reviewer: "any risk" seems off

Reviewer 2: Mitigate means lessen, so it is certainly not saying all risk is eradicated. But perhaps s/any risk/this risk would be an improvement.

  with a transaction that's timelocked and can't be confirmed until some
  point in the future.  As a result of this bribe, the miner won't
  confirm any transaction that could be mined immediately but which
  conflicts with the bribe (as long as the bribe pays a sufficiently
  higher feerate than any transactions it blocks).

    If this theory is accurate, it affects Hash TimeLock Contracts
    (HTLCs) which can be settled immediately by one party (Alice) using
    a preimage, or which can be settled later by a second party (Bob)
    after a timeout.  According to the authors, if Bob sees Alice use
    the preimage to spend the HTLC, Bob should send his conflicting
    timeout settlement transaction to miners with a sufficiently higher
    feerate than Alice's preimage settlement---even though miners can't
    immediately include Bob's transaction---bribing them to ignore
    Alice's transaction in favor of waiting to confirm his alternative
    transaction.  This allows Bob to steal the money Alice should
    receive.  HTLCs are currently used in LN, atomic swaps, and several
    other contract protocols.

    The authors of the paper propose a solution they call Mutually

>     Assured Destruction HTLCs (MAD-HTLCs).  These require Bob provide a
>     hashlocked bond that make him reveal his own preimage when sending

Reviewer 3: nit
```suggestion
    Assured Destruction HTLCs (MAD-HTLCs).  These require Bob to provide a
    hashlocked bond that makes him reveal his own preimage when sending
```

    his timeout settlement.  If both Alice and Bob reveal their
    respective preimages, miners will be able to claim both the
    payment/refund amount and the bond collateral---destroying the
    incentive for Bob to attempt stealing by bribing miners.

>     Downsides to this approach are that it requires Bob use more

Reviewer 3: nit
```suggestion
    Downsides to this approach are that it requires Bob to use more
```

  mailing list about creating software and a protocol for allowing
  applications to store presigned transactions with third-party services
  for later broadcast, perhaps with the software determining when to
  broadcast the transactions based on certain conditions being met.  This
  could be useful for software such as [vaults][topic vaults] and
  [watchtowers][topic watchtowers].  Swambo asks anyone interested in
  the idea, especially those interested in using it with their protocol, to
  contact him.

## Recently transcribed talks and conversations

*[Bitcoin Transcripts][] is the home for transcripts of technical
Bitcoin presentations and discussions. In this monthly feature, we
highlight a selection of the transcripts from the previous month.*

- **Magical Bitcoin:** Alekos Filini presented at LA BitDevs on [Magical
  Bitcoin][], an in-development set of tools and libraries for onchain
  wallet developers. He explained that coin selection logic and
  transaction signing logic currently have to be rewritten multiple times
  across multiple projects. Magical Bitcoin aims to provide peer
  reviewed, modular, and extensible components to address this. A longer

>   term ambition would be to provide a standard platform to build native

Reviewer 2: nit: s/standard/de facto standard

I initially had "de facto standard" and I think it makes it clearer. Without "de facto" the use of "standard" may be confused with "average or normal" rather than the intended meaning of "community approved and widely used as the go to platform"

Author: How about s/standard/popular/?  I think "community approved and widely used as the go to platform" is a lot of meaning to stuff into "de facto" (and I try to avoid foreign languages; Latin expressions like "de facto" may be used in many European languages, but be meaningless to people with other native languages; though, to be fair, I do allow myself to use Latin-derived abbreviations such as e.g., i.e., and etc.).

Reviewer 2: I don't like popular either.. too gauche ;) Maybe just get rid of "standard" altogether.

 "provide a ~~standard~~ platform ~~to~~ **for** ~~build~~ **building** native wallets and ~~integrate~~ **integrating** wallets into existing projects"

  compliant with the proposed BOLT13 (see [Newsletter #75][news75
  watchtower bolt]).  Delgado also highlighted how watchtowers are
  becoming increasingly critical in multiple settings such as Bitcoin
  [vault][topic vaults] designs, statechains, and atomic swaps in addition to LN.
  ([transcript][watchtowers ts], [video][watchtowers vid],
  [slides][watchtowers slides])

- **CoinSwap:** Aviv Milner presented on CoinSwap at the Wasabi Research
  Club. He explained the property of covertness and how Chris Belcher’s
  CoinSwap proposal (see [Newsletter #100][news100 coinswap]) provides
  covertness in a manner that other privacy schemes such as
  [coinjoin][topic coinjoin] fail to do. Milner also went through the
  motivation for routing CoinSwaps, namely to address the concern of
  unwittingly entering into a CoinSwap with an adversary such a chain
  surveillance company.  ([transcript][coinswap ts], [video][coinswap
  vid])

- **Taproot and schnorr multisignatures:** A wide ranging discussion at
  London Bitcoin Devs on the [BIP340][] specification of [schnorr
  signatures][topic schnorr signatures] with some of the co-authors such as

>   Tim Ruffing, Pieter Wuille, and Jonas Nick.  This covered some of the

Author: Just for the record, this was totally my fault.  I re-arranged the order of @michaelfolkson's two paragraphs in the section (to put them in temporal order).

    threshold signature schemes using schnorr signatures. He highlighted
    how the rogue key attack and an attack using Wagner’s algorithm (see
    [Newsletter #74][news74 wagner]) need to be addressed when designing
    secure multisignature schemes.  Ruffing has been working with
    collaborators on a speculative [MuSig][topic musig] signature scheme
    that only requires two rounds of interaction rather than three
    (without relying on zero knowledge proofs). Threshold signature
    schemes generally present further challenges such as requiring an
    honest majority, a broadcast channel, and careful consideration of
    parallel signing sessions.  ([transcript][ruffing ts],
    [video][ruffing vid], [slides][ruffing slides])

- **Sydney meetup discussion:** A number of Bitcoin developers and LN
  developers joined this Sydney meetup to discuss various topics. Ruben
  Somsen gave a short presentation on Succinct Atomic Swaps (SAS), see
  [Newsletter #98][news98 sas].  Somsen explained how SAS compared to
  Chris Belcher’s CoinSwap proposal, whether atomic swaps via privacy
  focused altcoins is a workable approach to obtain additional privacy,

>   and how SAS only requires one of the two block chains to have timelocks

Reviewer 2: nit: s/block chains/blockchains (it appears from doing a grep that you use both interchangeably on Optech but I think blockchains is much more widely used outside of Optech)

Author: "block chain" is what's in our style guide: https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#compound-words

I just checked the grep myself and most of them seem to be references to the Stanford Blockchain Conference or urls.

  and scripting opcodes.

    Other non-related topics included reliance on DNS seeds in Bitcoin
    Core, whether they present a viable attack vector against new full

>     nodes seeking to locate network peers for the very first time, and

Reviewer 2: nit: I think "and **how** work by Matt Corallo and Antoine Riard on downloading block headers over HTTP and allowing other alternative transports ~~that~~ could potentially mitigate this risk" reads better.

Please also note that this work was brought up briefly as possible mitigations but not discussed in any depth.

    that Bitcoin Core uses more bandwidth than necessary when it
    encounters segwit transactions that are invalid because of
    accidental mistakes.  It may also complicate the development of new
    relay protocols such as [package relay][topic package relay].

    The long-term solution to the problem is that transactions should be
    announced using a hash that commits to the entire transaction---both
    the legacy fields and the new segwit fields.  That's exactly what
    wtxids do, and they're already needed in the protocol to construct
    the witness merkle root in each block's coinbase transaction.  This
    new BIP proposes updating the P2P protocol's `inv` message that
    announces new transactions, and the `getdata` message that requests
    a transaction, to allow them both to use wtxids.  That will allow
    nodes to skip re-downloading a transaction if it has the same wtxid
    as a transaction that was previously found to be invalid or
    unacceptable.

    The BIP339 proposal also adds the additional feature negotiation
    between newly started nodes described in [Newsletter #87][news87

>     negotiation].

Reviewer 3: Is there value in mentioning the [implementation](http://github.com/bitcoin/bitcoin/pull/18044)? (I'm aware you can also get to it from the BIP)

  [payjoin][topic payjoin] protocol originally implemented in BTCPay
  (see [Newsletter #94][news94 btcpayjoin]).  Payjoin allows a spender
  and a receiver to both add UTXOs to a transaction.  This decreases the
  reliability of the assumption made by third-party block chain
  surveillance companies that any set of UTXOs spent in the same
  transaction were all received by the same person.  BIP78 is based on
  the [BIP79][] specification of the Bustapay payjoin variant (see
  [Newsletter #13][news13 bustapay]) but contains several notable
  differences, including different extensions to [BIP21][] `bitcoin:`
  URIs, usage of PSBTs, and some additional requirements designed to
  enhance privacy.  Several programs already support this protocol and
  several more are currently working on adding support.  The PR for this
  BIP received significant discussion and may provide useful reference
  material for anyone interested in the subject.

- [BIPs #550][] revises the [BIP8][] alternative to [BIP9][]
  versionbits-based soft fork deployment.  BIP8 previously allowed a
  soft fork to be activated by miner signaling using the same parameters

>   as would be used for BIP9, but it required that the soft fork activate

Reviewer 3: ```suggestion
  as would be used for BIP9, but it required that the soft fork activates
```

Author: I think "activate" matches the tense of "required".

Reviewer 4: 'activate' is correct. Nice use of the subjunctive mood @harding  :mage_man:

  at the end of the signaling period even if miners were still not
  signaling readiness to follow the new consensus rules.  This could be
  used to override miners who were obstructing activation of a fork, but
  it could also lead to diverging block chains between nodes that
  enforced the fork's new consensus rules and those that didn't.

>     The main changes to BIP8 from its previous version is that it

Reviewer 3: ```suggestion
    The main change to BIP8 from its previous version is that it
```

[watchtowers vid]: https://www.youtube.com/watch?v=Vkq9CVxMclE
[watchtowers slides]: https://srgi.me/resources/slides/Potzblitz!2020-Watchtowers.pdf
[coinswap ts]: https://diyhpl.us/wiki/transcripts/wasabi-research-club/2020-06-15-coinswap/
[coinswap vid]: https://www.youtube.com/watch?v=Pqz7_Eqw9jM
[socratic ts]: https://diyhpl.us/wiki/transcripts/london-bitcoin-devs/2020-06-16-socratic-seminar-bip-schnorr/
[ruffing ts]: https://diyhpl.us/wiki/transcripts/london-bitcoin-devs/2020-06-17-tim-ruffing-schnorr-multisig/
[socratic vid]: https://www.youtube.com/watch?v=uE3lLsf38O4
[ruffing vid]: https://www.youtube.com/watch?v=8Op0Glp9Eoo
[ruffing slides]: https://slides.com/real-or-random/taproot-and-schnorr-multisig
[sydney ts]: https://diyhpl.us/wiki/transcripts/sydney-bitcoin-meetup/2020-06-23-socratic-seminar/
[magical bitcoin]: https://magicalbitcoin.org/
[news75 watchtower bolt]: /en/newsletters/2019/12/04/#proposed-watchtower-bolt
[news98 sas]: /en/newsletters/2020/05/20/#two-transaction-cross-chain-atomic-swap-or-same-chain-coinswap
[teos]: https://github.com/talaia-labs/python-teos

> [news74 wagner]: http://devbcop/en/newsletters/2019/11/27/#schnorr-taproot-updates

Reviewer 3: ```suggestion
[news74 wagner]: /en/newsletters/2019/11/27/#schnorr-taproot-updates
```

> [news100 coinswap]: http://devbcop/en/newsletters/2020/06/03/#design-for-a-coinswap-implementation

Reviewer 3: ```suggestion
[news100 coinswap]: /en/newsletters/2020/06/03/#design-for-a-coinswap-implementation
```

    Alice's transaction in favor of waiting to confirm his alternative
    transaction.  This allows Bob to steal the money Alice should
    receive.  HTLCs are currently used in LN, atomic swaps, and several
    other contract protocols.

    The authors of the paper propose a solution they call Mutually
    Assured Destruction HTLCs (MAD-HTLCs).  These require Bob to provide a
    hashlocked bond that makes him reveal his own preimage when sending
    his timeout settlement.  If both Alice and Bob reveal their
    respective preimages, miners will be able to claim both the
    payment/refund amount and the bond collateral---destroying the
    incentive for Bob to attempt stealing by bribing miners.

    Downsides to this approach are that it requires Bob to use more
    collateral---raising the cost of using HTLCs---and it uses more
    block chain space when MAD-HTLCs are settled onchain compared to
    traditional HTLCs---also raising costs.  It was also
    [claimed][harding mad miner] in the mailing list discussion that
    MAD-HTLCs might have their own problems with theft when the
    bonded user's counterparty is a miner (e.g. Alice is a miner).

>     ZmnSCPxj [noted][zmnscpxj scorched earth] that there already exists

Reviewer: that a mechanism already exists

  [presentation video][ruffing vid], [presentation slides][ruffing
  slides])

- **Sydney meetup discussion:** A number of Bitcoin and LN developers
  joined this Sydney meetup to discuss various topics. Ruben Somsen gave
  a short presentation on Succinct Atomic Swaps (SAS) (see [Newsletter
  #98][news98 sas]) and how it compares to Chris Belcher’s coinswap
  proposal.  Another topic was Bitcoin Core's reliance on DNS seeds to
  find initial peers, whether it's a potential attack vector against newly
  started nodes, and how work by Matt Corallo and Antoine Riard on
  allowing [alternative transports][Bitcoin Core #18988] could help
  mitigate some risks.  Finally, Lloyd Fournier discussed how
  experimenting with his toy Rust implementation of secp256k1
  (secp256kfun) led to a small [fix][libsecp256k1 #732] in the ECDSA
  signature code in the actual secp256k1 library. The transcript was
  anonymized to encourage open discussion.  ([transcript][sydney ts])

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

> - [HWI 1.1.2][hwi 1.1.2]: this release includes support

Reviewer 5: Should we note this is a RC?

Author: FIxed, thanks!

Reviewer: > Should we note this is a RC?

I could be wrong but it looks like the RC was a week ago and this is a release: https://github.com/bitcoin-core/HWI/tags

Reviewer 5: @jonatack Yep: https://github.com/bitcoin-core/HWI/releases/tag/1.1.2

  for handling [PSBTs][topic psbt] with serialized previous segwit
  transactions, which is now required or suggested for several hardware
  wallets in response to the [fee overpayment attack][news101 fee
  overpayment].

- [LND 0.10.2-beta.rc2][lnd 0.10.2-beta]: this release candidate for an
  LND maintenance release is now available for testing.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],

> [Hardware Wallet Interface][hwi], [Bitcoin Improvement Proposals

Reviewer: perhaps "Hardware Wallet Interface (HWI)" to begin introducing the acronym

  of 2020) and break C++11 compatibility later in 0.22 (expected mid-2021).
  Although this change won't affect most users, those planning to build
  the 0.22 release on older systems may want to check their toolchain's standard
  support against the [C++ compiler support matrix][cpp compiler support].

- [Bitcoin Core #11413][] updates the `bumpfee`, `fundrawtransaction`,
  `sendmany`, `sendtoaddress`, and `walletcreatefundedpsbt` RPCs to
  allow manually specifying the feerate to use in the newly created or
  updated transaction.  By default, these commands still use either an
  automatic feerate computed by Bitcoin Core's built-in transaction fee
  estimate or the fallback feerate if there's not enough data for
  estimation.  For details on how to use the updated RPCs, see the
  [proposed release note][bcc11413rn].  Despite not affecting any major
  system, this PR was open for almost three years---the second longest
  of any PR that has been merged into Bitcoin Core so far.  We thank the
  author, Kalle Alm, and everyone else involved for their persistence.

- [Eclair #1466][] decreases the length of time that an Eclair node will wait

>   before closing a channel when a peer becomings unresponsive while an HTLC is

Reviewer: becomes

  pending. To prevent payments from becoming stuck forever, each HTLC includes
  a timelock. If the timelock expires, then the sending party can reclaim
  the funds in the HTLC. To prevent the sending party from pulling back
  funds that an intermediate party has already paid to the next hop, the
  intermediate party must settle the pending HTLC onchain if the sending
  party becomes unresponsive. This PR increases Eclair's safety window---it

>   will now broadcast the on chain transaction 24 blocks before the HTLC

Reviewer: onchain

Reviewer: perhaps end the sentence at "6" or "instead of the previous value of six."

Should we mention something along the lines that this change effectively lowers `expiry-delta-blocks`, so Eclair users may wish to increase it.

Author: I think we'll skip mentioning that to avoid having to add extra information about the relationship between the parameters.  Hopefully Eclair will put something about that into their release notes when they release this change.

  timeout, instead of 6 blocks before the HTLC timeout.

- [LND #4018][] adds the ability to delay forwarding a payment (HTLC),
  giving an external process the ability to review it and decide whether
  to cancel it or continue forwarding it.  This is similar to [hold
  invoices][topic hold invoices] but it applies to payments being routed
  to a node's peers rather than received by the node itself.  Several
  use cases are described in the PR, such as being able to hold the
  HTLC, detect that its next hop is offline (e.g. a mobile node), send a

>   notification to that node which will restart it, and then continue

Reviewer: perhaps `s/, send/, or send/` combined with `s/it, and/it and/` to help readers understand that from this point we are describing a single last use case (if I'm understanding the writing correctly).

  relaying the HTLC.

- [LND #4106][] adds a `getrecoveryinfo` RPC that returns information
  about the progress of restoring from a backup.

- [BIPs #933][] adds the [BIP339][] specification for transaction
  relay announcement using Witness Transaction Identifiers (wtxids).
  Currently, nodes announce new unconfirmed transactions to their peers
  using the transaction's txid, which is a hash over the legacy fields
  of the transaction.  The new fields used in segwit transactions are
  not included in the txid, which was necessary to eliminate third-party
  and counterparty transaction malleability.  However, during the review

>   of segwit, Peter Todd [noticed][todd relay malleability] that

Reviewer: perhaps mention when he noticed this here (it's only indirectly hinted at two paragraphs down)

  changing its txid.

    Before Todd's discovery, this was a problem: Bitcoin Core would
    track the txids of invalid transactions it had recently seen and
    refuse to download them again.  That meant a malicious node could
    prevent any of its peers from downloading a valid segwit transaction
    from any of their other peers by sending them a mutated version of
    that transaction.  Arbitrary transaction censorship, even just at
    the relay level, is bad by itself---but it has especially severe
    consequences for time-sensitive protocols such as LN.

    At the time of Todd's analysis, segwit was nearing release, so a
    [quick fix][Bitcoin Core #8312] was implemented that prevents
    Bitcoin Core from caching the rejection status of segwit
    transactions that fail for the type of witness-related errors that
    malicious nodes can inject.  This eliminated the issue, but it means
    that Bitcoin Core uses more bandwidth than necessary when it
    encounters segwit transactions that are invalid because of
    accidental mistakes.  It may also complicate the development of new
    relay protocols such as [package relay][topic package relay].

>     The long-term solution to the problem is that transactions should be

Reviewer: "long-term" is used here (and "third-party is used thrice in this issue) but "long term", "medium term", "short term", and also "longer term" are used later on.

    announced using a hash that commits to the entire transaction---both
    the legacy fields and the new segwit fields.  That's exactly what
    wtxids do, and they're already needed in the protocol to construct
    the witness merkle root in each block's coinbase transaction.  This
    new BIP proposes updating the P2P protocol's `inv` message that
    announces new transactions, and the `getdata` message that requests
    a transaction, to allow them both to use wtxids.  That will allow
    nodes to skip re-downloading a transaction if it has the same wtxid
    as a transaction that was previously found to be invalid or
    unacceptable.

    The BIP339 proposal also adds the additional feature negotiation
    between newly started nodes described in [Newsletter #87][news87
    negotiation].  See also the [proposed implementation][Bitcoin Core

>     #18044] for Bitcoin Core.

Reviewer: Not worth linking to, but I'll mention that there are two review club resources on wtxid relay (https://bitcoincore.reviews/18044) and package relay (https://bitcoincore.reviews/16401).

  [Newsletter #13][news13 bustapay]) but contains several notable
  differences, including different extensions to [BIP21][] `bitcoin:`
  URIs, usage of PSBTs, and some additional requirements designed to
  enhance privacy.  Several programs already support this protocol and
  several more are currently working on adding support.  The PR for this
  BIP received significant discussion and may provide useful reference
  material for anyone interested in the subject.

- [BIPs #550][] revises the [BIP8][] alternative to [BIP9][]
  versionbits-based soft fork deployment.  BIP8 previously allowed a
  soft fork to be activated by miner signaling using the same parameters
  as would be used for BIP9, but it required that the soft fork activate
  at the end of the signaling period even if miners were still not
  signaling readiness to follow the new consensus rules.  This could be
  used to override miners who were obstructing activation of a fork, but
  it could also lead to diverging block chains between nodes that
  enforced the fork's new consensus rules and those that didn't.

    The main change to BIP8 from its previous version is that it

>     may now be used initially without requiring mandatory

Reviewer: initially used?

Author: I think this is the best way to say it, but that's just my instinct. :man_shrugging:

>     lockin.  Implementations may choose to commit to lockin after their

Reviewer: several occurences: s/lockin/lock-in/

    initial deployment of a BIP8-activated soft fork.  BIP8 mandates
    signaling for a period after the mandatory lockin height, which will

>     trigger even deployments without mandatory lockin to begin

Reviewer: "even" seems like it should be before "trigger"

[coinswap vid]: https://www.youtube.com/watch?v=Pqz7_Eqw9jM
[socratic ts]: https://diyhpl.us/wiki/transcripts/london-bitcoin-devs/2020-06-16-socratic-seminar-bip-schnorr/
[ruffing ts]: https://diyhpl.us/wiki/transcripts/london-bitcoin-devs/2020-06-17-tim-ruffing-schnorr-multisig/
[socratic vid]: https://www.youtube.com/watch?v=uE3lLsf38O4
[ruffing vid]: https://www.youtube.com/watch?v=8Op0Glp9Eoo
[ruffing slides]: https://slides.com/real-or-random/taproot-and-schnorr-multisig
[sydney ts]: https://diyhpl.us/wiki/transcripts/sydney-bitcoin-meetup/2020-06-23-socratic-seminar/
[magical bitcoin]: https://magicalbitcoin.org/
[news75 watchtower bolt]: /en/newsletters/2019/12/04/#proposed-watchtower-bolt
[news98 sas]: /en/newsletters/2020/05/20/#two-transaction-cross-chain-atomic-swap-or-same-chain-coinswap
[teos]: https://github.com/talaia-labs/python-teos
[news74 wagner]: /en/newsletters/2019/11/27/#schnorr-taproot-updates
[news100 coinswap]: /en/newsletters/2020/06/03/#design-for-a-coinswap-implementation
[cpp compiler support]: https://en.cppreference.com/w/cpp/compiler_support#C.2B.2B17_features
[potzblitz]: https://www.youtube.com/playlist?list=PLwgam6YBS0-jk1TlXD7QXDjTYJh-eJn_X

> [knw paper]: https://arxiv.org/pdf/2006.08513.pdf

Reviewer 4: Incorrect link. Should be https://eprint.iacr.org/2020/774

Author: Whoops, too many tabs open with papers to read!

---
title: 'Bitcoin Optech Newsletter #104'
permalink: /en/newsletters/2020/07/01/
name: 2020-07-01-newsletter
slug: 2020-07-01-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter summarizes a discussion about mining incentives
related to HTLCs and links to an announcement about a proposed service to
store and relay presigned transactions.  Also included are our regular
sections with recently transcribed talks and conversations, new releases
and release candidates, and notable changes to popular Bitcoin
infrastructure projects.

## Action items

*None this week.*

## News

{% comment %}<!-- FIXME Reviewers: the following item links to two of my
posts; please be extra critical of those links and their related
content.  I'll respect any decisions you make regarding them -->{% endcomment %}

> - **Discussion of HTLC mining incentives:** Itay Tsabary, Matan

Reviewer 4: Very long initial sentence. Consider splitting.

    [claimed][harding mad miner] in the mailing list discussion that
    MAD-HTLCs might have their own problems with theft when the
    bonded user's counterparty is a miner (e.g. Alice is a miner).

    ZmnSCPxj [noted][zmnscpxj scorched earth] that there already exists
    a mechanism to allow Alice to discourage Bob from attempting his
    theft: Alice can enact a [scorched earth][] policy where she spends
    all of her legitimate funds to fees in order to prevent Bob from
    receiving them.  In theory, this should prevent Bob from even trying
    the theft.  It was also [explained][harding myopic] that nearly all
    miners would need to switch to the proposed bribe-accepting software
    before the attack would become particularly effective.

    In the short term, and probably the medium term, this attack does
    not appear to pose any significant danger to HTLCs whose receivers
    resolve their preimages well before their timelocks expire.  In the
    long term, this is an incentive compatibility concern that
    protocol developers may want to keep in mind.

- **Proposed service for storing, relaying, and broadcasting presigned transactions:**

>   Jacob Swambo sent a [request for comments][swambo rfc] to the Bitcoin-Dev

Reviewer 4: Again, consider splitting this very long sentence.

  mailing list about creating software and a protocol for allowing
  applications to store presigned transactions with third-party services
  for later broadcast, perhaps with the software determining when to
  broadcast the transactions based on certain conditions being met.  This
  could be useful for software such as [vaults][topic vaults] and
  [watchtowers][topic watchtowers].  Swambo asks anyone interested in
  the idea, especially those interested in using it with their protocol, to
  contact him.

## Recently transcribed talks and conversations

*[Bitcoin Transcripts][] is the home for transcripts of technical
Bitcoin presentations and discussions. In this monthly feature, we
highlight a selection of the transcripts from the previous month.*

- **Magical Bitcoin:** Alekos Filini presented at LA BitDevs on [Magical
  Bitcoin][], an in-development set of tools and libraries for onchain
  wallet developers. He explained that coin selection logic and
  transaction signing logic currently have to be rewritten multiple times

>   across multiple projects. Magical Bitcoin aims to provide peer

Reviewer 4: should this be "peer-reviewed" ?

>   reviewed, modular, and extensible components to address this. A longer

Reviewer 4: s/address this/address these needs/ ? (What does 'this' refer to here?)

Reviewer 2: 'This' is the problem of coin selection logic and transaction signing logic having to be rewritten across multiple projects rather than just written and maintained once for Magical Bitcoin and then reused across multiple projects. So addressing this inefficiency and avoiding replicated work.

>   term ambition would be to provide a platform for building native

Reviewer 4: s/would be/is

  wallets and integrating them with existing projects.  Filini demoed
  the current functionality, which includes a playground with a Policy
  to [Miniscript][topic miniscript] compiler and some rudimentary
  visualizations. It is written in Rust and leverages the
  rust-miniscript library and the open source Esplora block explorer.
  ([transcript][mb ts], [video][mb vid])

> - **Watchtowers and BOLT13:** Sergi Delgado appeared on Potzblitz to

Reviewer 4: I'd never heard of Potzblitz before. Perhaps link to https://www.youtube.com/playlist?list=PLwgam6YBS0-jk1TlXD7QXDjTYJh-eJn_X ?

  discuss the latest state of [watchtower][topic watchtowers]
  development and a proposed watchtower protocol specification. He

>   explored the various trade-offs when designing a watchtower and the

Reviewer 4: tradeoff seems more common in the repo than trade-off. Not sure which is preferred.

  fees charged.  Delgado is working on the watchtower implementation
  [The Eye of Satoshi][teos] at Talaia Labs, which is aiming to be
  compliant with the proposed BOLT13 (see [Newsletter #75][news75
  watchtower bolt]).  Delgado also highlighted how watchtowers are
  becoming increasingly critical in multiple settings such as Bitcoin
  [vault][topic vaults] designs, statechains, and atomic swaps in addition to LN.
  ([transcript][watchtowers ts], [video][watchtowers vid],
  [slides][watchtowers slides])

- **CoinSwap:** Aviv Milner presented on CoinSwap at the Wasabi Research
  Club. He explained the property of covertness and how Chris Belcher’s
  CoinSwap proposal (see [Newsletter #100][news100 coinswap]) provides
  covertness in a manner that other privacy schemes such as
  [coinjoin][topic coinjoin] fail to do. Milner also went through the
  motivation for routing CoinSwaps, namely to address the concern of
  unwittingly entering into a CoinSwap with an adversary such a chain
  surveillance company.  ([transcript][coinswap ts], [video][coinswap
  vid])

- **Taproot and schnorr multisignatures:** Some of the co-authors of [BIP340][],

>   including Tim Ruffing, Pieter Wuille, and Jonas Nick, held a wide ranging

Reviewer 4: s/wide ranging/wide-ranging/ (even better, just drop "wide ranging")

  discussion at London Bitcoin Devs on the specification of [schnorr
  signatures][topic schnorr signatures].  This covered some of the

>   earlier ideas for implementing schnorr in Bitcoin and what the

Reviewer 4: s/implementing schnorr/implementing schnorr signatures/

  co-authors thought the community should be concerned about when
  considering a possible future soft fork deployment. Wuille expressed
  his view that it is unlikely at this stage that there are bugs in the
  proposed consensus implementation and that he is more concerned with
  how schnorr is adopted and what is built using it.
  ([transcript][socratic ts], [video][socratic vid])

    The following day, Tim Ruffing presented on multisignature and
    threshold signature schemes using schnorr signatures. He highlighted
    how the rogue key attack and an attack using Wagner’s algorithm (see
    [Newsletter #74][news74 wagner]) need to be addressed when designing
    secure multisignature schemes.  Ruffing has been working with
    collaborators on a speculative [MuSig][topic musig] signature scheme
    that only requires two rounds of interaction rather than three

>     (without relying on zero knowledge proofs). Threshold signature

Reviewer 4: This shouldn't be in a parenthetical. Better would be "..rather than three, without relying..."

    schemes generally present further challenges such as requiring an
    honest majority, a broadcast channel, and careful consideration of
    parallel signing sessions.  ([transcript][ruffing ts],
    [video][ruffing vid], [slides][ruffing slides])

> - **Sydney meetup discussion:** A number of Bitcoin developers and LN

Reviewer 4: I suggest we limit this section to 4 transcripts. If you agree, I think we should drop this one, since it's the least focused of the 5 talks.

Reviewer 2: Not my call to make but let me know for next month what the requirements are please. Always open to feedback on whether it is too long/too short/not of interest to Optech readers etc. I certainly thought there should be a longer section on the Schnorr multisig stuff this month.

Author: In our initial conversations about the section, I told Michael, "I'd prefer the section be no more than 5 transcripts and a maximum of 700 words".   We're currently at 682 words for the section (including link anchors), so if >4 transcripts is a problem, I want to be clear that's my fault.

I think maybe this section looks so long because it has two items that each have two paragraphs.  Maybe instead of a 4 or 5 item limit, we have a 5 paragraph total limit?  Also, we could lower the maximum word count limit a bit to make the section shorter no matter what the item/paragraph limit is.

I'll make the other edits you suggest in one commit, and then---in a separate commit---I'll see if I can't compress this down to just five paragraphs without losing too much information (or creating crazy long paragraphs).  Would that be a reasonable compromise for this month at least?

    (HTLCs) which can be settled immediately by one party (Alice) using
    a preimage, or which can be settled later by a second party (Bob)
    after a timeout.  According to the authors, if Bob sees Alice use
    the preimage to spend the HTLC, Bob should send his conflicting
    timeout settlement transaction to miners with a sufficiently higher
    feerate than Alice's preimage settlement---even though miners can't
    immediately include Bob's transaction---bribing them to ignore
    Alice's transaction in favor of waiting to confirm his alternative
    transaction.  This allows Bob to steal the money Alice should
    receive.  HTLCs are currently used in LN, atomic swaps, and several
    other contract protocols.

    The authors of the paper propose a solution they call Mutually
    Assured Destruction HTLCs (MAD-HTLCs).  These require Bob to provide a
    hashlocked bond that makes him reveal his own preimage when sending
    his timeout settlement.  If both Alice and Bob reveal their
    respective preimages, miners will be able to claim both the
    payment/refund amount and the bond collateral---destroying the

>     incentive for Bob to attempt stealing by bribing miners.

Reviewer: s/stealing/to steal/ (I think)

>     Downsides to this approach are that it requires Bob to use more

Reviewer: s/Downsides/The downsides/

    collateral---raising the cost of using HTLCs---and it uses more

>     block chain space when MAD-HTLCs are settled onchain compared to

Reviewer: (I know it's in the style guide but do find that "onchain" looks like a misspelled version of on-chain)

Author: I have no strong preference either way.  If you want us to change it, feel free to open a PR updating the style guide.

    traditional HTLCs---also raising costs.  It was also

>     [claimed][harding mad miner] in the mailing list discussion that

Reviewer: I understand why the author here is not named here and at the end of this paragraph, but it might be valuable to readers to add "by this author" just as other valued participants would be named. Feel free to ignore; I don't want to waste your time :)

Author: You're quite likely right that it would be valuable to readers.  However, you can be sure that as long as I have a say in it, anywhere there's a spotlight, it'll be pointed as far away as possible from me.  :-)

    over HTTP and allowing other [alternative transports][Bitcoin Core
    #18988] could potentially mitigate this risk. Lloyd Fournier
    also discussed how experimenting with his toy Rust implementation of
    secp256k1 (secp256kfun) led to a small [fix][libsecp256k1 #732] in the ECDSA signature
    code in the actual secp256k1 library. The transcript was anonymized
    to encourage open discussion.  ([transcript][sydney ts])

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [HWI 1.1.2-rc.2][hwi 1.1.2]: this release candidate includes support
  for handling [PSBTs][topic psbt] with serialized previous segwit
  transactions, which is now required or suggested for several hardware
  wallets in response to the [fee overpayment attack][news101 fee
  overpayment].  Experienced users, and developers using HWI in their own
  applications, may wish to test whether this prerelease resolves the
  issues related to recent hardware wallet firmware updates.

- [LND 0.10.2-beta.rc2][lnd 0.10.2-beta]: this release candidate for an

>   LND maintenance release is now available for testing.

Reviewer: One project to consider covering in terms of releases might be JoinMarket. Mentioning it because waxwing is announcing a release: https://x0f.org/@waxwing/104428846061585630

Author: I'd love to cover JoinMarket, but I think it'd be hard for us to start covering it without also covering Wasabi and maybe Samourai Wallet out of fairness.  Although it's not a hard rule, currently I try to only use this section and the notable merges section for infrastructure projects---software that's often used in other software or services.

Reviewer: I agree, infra is a good distinction/place to draw the line, and it's good to say it here in the intro to this section. I wonder if it should be clarified in the opening summary where the term is already present.
